Add the callable proposal. (#1009)
apple/swift-evolution: proposals/0252-callable.md

        return result
    }
}

let polynomial = Polynomial(coefficients: [2, 3, 4])
print(polynomial.evaluated(at: 2)) // => 24
```

The mathematical notation for function application is simply `output = f(input)`. Using subscript methods achieve a similar application syntax `f[x]`, but subscripts and square brackets typically connote "indexing into a collection", which is not the behavior here.

```swift
extension Polynomial {
    subscript(input: Float) -> Float {
        ...
    }
}
let polynomial = Polynomial(coefficients: [2, 3, 4])
// Subscript syntax, may be confusing.
print(polynomial[2]) // => 24
```

The proposed feature enables the same call syntax as the mathematical notation:
```swift
extension Polynomial {
    call(_ input: Float) -> Float {
        ...
    }
}
let polynomial = Polynomial(coefficients: [2, 3, 4])
// Callable syntax.
print(polynomial(2)) // => 24
```

#### Bound closures

Variable-capturing closures can be modeled explicitly as structs that store the bound variables. This representation is more performant and avoids the type-erasure of closure contexts.

```swift
// Represents a nullary function capturing a value of type `T`.
struct BoundClosure<T> {

>   var function: (T) -> Void

Reviewer: 4 space indentation

In a nutshell, we propose to introduce a new declaration syntax with the keyword `call` :

```swift
struct Adder {
    var base: Int
    call(_ x: Int) -> Int {
        return base + x
    }
}
```

Values that have a `call` member can be applied like functions, forwarding arguments to the `call` member.

```swift
let add3 = Adder(base: 3)
add3(10) // => 13
```

## Motivation

Currently, in Swift, only a few kinds of values are syntactically callable:
- Values with function types.
- Type names (e.g. `T` can be called like `T(...)`, which is desugared to `T.init(...)`).
- Values with a `@dynamicCallable` type.

However, callable syntax can also be useful for other values, primarily those that behave like functions. This includes:
- Values that represent functions: mathematical functions, function expressions, etc.
- Values that have one main use and want to provide a simple callable interface: neural network layers, parsers, efficient bound closures, etc.

Here are some concrete sources of motivation.

### Values representing functions

> Values of some types exactly represent functions: in the mathematical sense (a mapping from inputs to outputs), or in the context of programming languages.

Reviewer: `Values of some types exactly represent functions` is a bit confusing, particularly `some types`.
```suggestion
Some types exactly represent functions: in the mathematical sense (a mapping from inputs to outputs), or in the context of programming languages.
```

Author: "Types representing functions" is incorrect. I intentionally changed the original "some types ..." to "values of some types".

Functions are values, not types. Thus types cannot represent functions.

Reviewer: How about `Some values exactly represent functions?`

Author: That is technically correct, but it is correct more because "some values" can just _be_ functions, which feels like tautology and doesn't go into more details than the section title.
I changed it to "Values of some nominal types exactly represent functions".

```swift
extension Polynomial {
    subscript(input: Float) -> Float {
        ...
    }
}
let polynomial = Polynomial(coefficients: [2, 3, 4])
// Subscript syntax, may be confusing.
print(polynomial[2]) // => 24
```

The proposed feature enables the same call syntax as the mathematical notation:
```swift
extension Polynomial {
    call(_ input: Float) -> Float {
        ...
    }
}
let polynomial = Polynomial(coefficients: [2, 3, 4])
// Callable syntax.
print(polynomial(2)) // => 24
```

#### Bound closures

Variable-capturing closures can be modeled explicitly as structs that store the bound variables. This representation is more performant and avoids the type-erasure of closure contexts.

```swift
// Represents a nullary function capturing a value of type `T`.
struct BoundClosure<T> {
  var function: (T) -> Void
  var value: T

  call() { return function(value) }
}

let x = "Hello world!"
let closure = BoundClosure(function: { print($0) }, value: x)
closure() // prints "Hello world!"
```

> A callable syntactic sugar would enable `BoundClosure` instances to be applied like normal functions.

Reviewer: ```suggestion
A call syntax sugar would enable `BoundClosure` instances to be applied like normal functions.
```

Reviewer: Maybe hyphenize as `call-syntax sugar`.

One option is to use `func(...)` without an identifier name. Since the word "call" does not appear, it is less clear that this denotes a call-syntax delegate method. Additionally, it’s not clear how direct references would work: the proposed design of referencing `call` declarations via `foo.call` is clear and consistent with the behavior of `init` declarations.

To make unnamed `func(...)` less weird, one option is to add a `call` declaration modifier: `call func(...)`. The word `call` appears in both this option and the proposed design, clearly conveying "call-syntax delegate method". However, declaration modifiers are currently also treated as keywords, so with both approaches, parser changes to ensure source compatibility are necessary. `call func(...)` requires additional parser changes to allow `func` to sometimes not be followed by a name. The authors lean towards `call` declarations for terseness.

#### Use an attribute to mark call-syntax delegate methods

```swift
struct Adder {
    var base: Int
    @callDelegate
    func addingWithBase(_ x: Int) -> Int {
        return base + x
    }
}
```

> This approach achieves a similar effect as `call` declarations, except that methods can have a custom name and be directly referenced by that name. This is useful for types that want to make use of the callable syntactic sugar, but for which the name "call" does not accurately describe the callable functionality.

Reviewer: ```suggestion
This approach achieves a similar effect as `call` declarations, except that methods can have a custom name and be directly referenced by that name. This is useful for types that want to make use of the call syntax sugar, but for which the name "call" does not accurately describe the callable functionality.
```

* Proposal: SE-NNNN
* Authors: [Richard Wei](https://github.com/rxwei), [Dan Zheng](https://github.com/dan-zheng)
* Review Manager: TBD
* Status: Implementation in progress
* Implementation: https://github.com/apple/swift/pull/23517

## Introduction

Note: throughout this document, let "call-syntax" refer to the syntax of applying a function to arguments: `f(x, y, z)`.

This proposal introduces [callables](https://en.wikipedia.org/wiki/Callable_object) to Swift. Callables are values that define function-like behavior and can be applied using function application syntax.

In a nutshell, we propose to introduce a new declaration syntax with the keyword `call` :

```swift
struct Adder {
    var base: Int
    call(_ x: Int) -> Int {
        return base + x
    }
}
```

Values that have a `call` member can be applied like functions, forwarding arguments to the `call` member.

```swift
let add3 = Adder(base: 3)
add3(10) // => 13
```

## Motivation

Currently, in Swift, only a few kinds of values are syntactically callable:
- Values with function types.
- Type names (e.g. `T` can be called like `T(...)`, which is desugared to `T.init(...)`).
- Values with a `@dynamicCallable` type.

> However, callable syntax can also be useful for other values, primarily those that behave like functions. This includes:

Reviewer: Unify terminology: rename "callable syntax"and "call-syntax" to "call syntax"?
There are four mentions of "callable syntax".

Author: Done.
